### PR TITLE
Support bazelisk

### PR DESCRIPTION
**Pull request by zoidbergwill**
Closes https://github.com/bazelbuild/bazel-watcher/issues/339

TODO:
- [x] Add tests

### CODE REVIEW

File: bazel/bazel.go [hunk 1]
```diff
@@ -73,6 +91,11 @@ func findBazel() (string) {
 	if npmPath, err := bazelNpmPath(filepath.ToSlash(os.Args[0])); err == nil {
 		return filepath.FromSlash(npmPath)
 	}
+	// Frontend devs may have installed @bazel/bazelisk and @bazel/ibazel from npm
+	// If they also have bazelisk in the $PATH, we want to resolve this one, to avoid version skew
+	if npmPath, err := bazeliskNpmPath(filepath.ToSlash(os.Args[0])); err == nil {
```
Comment (achew22) [inline bazel/bazel.go]: You should probably check for Bazelisk before checking for Bazel.

File: bazel/bazel.go [hunk 2]
```diff
@@ -73,6 +91,11 @@ func findBazel() (string) {
...
+		return filepath.FromSlash(npmPath)
+	}
 
 	// Check in $PATH for system-installed Bazel
 	if path, err := exec.LookPath("bazel"); err == nil {
```
Comment (achew22) [inline bazel/bazel.go]: Can you also check to see if Bazelisk is on the path?

**Comment by zoidbergwill** [comment]
Thanks for the quick review @achew22 , those are great points. Updated.

File: bazel/bazel.go [hunk 3]
```diff
@@ -63,17 +63,43 @@ func bazelNpmPath(ibazelBinPath string) (string, error) {
...
+func bazeliskNpmPath(ibazelBinPath string) (string, error) {
+	s := strings.Split(ibazelBinPath, "/")
+	for i := 0; i + 4 < len(s); i++ {
+		prefix, nm, scope, pkg, dir, bin := s[0:i], s[i], s[i+1], s[i+2], s[i+3], s[i+4]
+		if nm == "node_modules" && scope == "@bazel" && pkg == "ibazel" && dir == "bin" {
+			name := strings.Join(append(prefix, nm, scope, "bazelisk", "bazelisk-" + bin), "/")
+			return name, nil
+		}
+	}
+	return "", errors.New("bazel binary not found in @bazel/bazel package")
```
Comment (zoidbergwill) [inline bazel/bazel.go]: bazelisk, silly

```suggestion
	return "", errors.New("bazel binary not found in @bazel/bazelisk package")
```

File: bazel/bazel.go [hunk 4]
```diff
@@ -63,17 +63,43 @@ func bazelNpmPath(ibazelBinPath string) (string, error) {
 	return "", errors.New("bazel binary not found in @bazel/bazel package")
 }
 
+// bazeliskNpmPath looks up a relative path to a binary from @bazel/bazelisk
+// This is used as an alternate resolution when no bazel binary is in the $PATH
+// When running from the @bazel/ibazel npm package, our binary is
+// /DIR/node_modules/@bazel/ibazel/bin/darwin_amd64/ibazel
+// We can find bazelisk in
+// /DIR/node_modules/@bazel/bazelisk/bazelizk-darwin_amd64
+func bazeliskNpmPath(ibazelBinPath string) (string, error) {
```
Comment (achew22) [inline bazel/bazel.go]: Could I have you add a test for this? It doesn't depend on any system calls or anything so a [table driven test](https://github.com/golang/go/wiki/TableDrivenTests) would be sufficient.

Comment (zoidbergwill) [reply bazel/bazel.go]: Sorry, added these earlier but couldn't figure out why I was getting a segfault, but fixed now.

Fortunately they caught me missing the `.exe` extension for Windows. I'm not 100% comfortable with how I'm doing the platform detection though.

**Pull request merged**